### ISSUE TITLE
Bug: flash attention produces corrupt output with batched generation

### ISSUE THREAD

**Issue by nytopop** [ggerganov/llama.cpp#7903]
### What happened?

When flash attention is enabled, generating parallel sequences in a batch produces corrupt output on some GPUs. It seems to only happen on certain batch sizes. I've tested with llama3 8b, mistral 7b, and qwen2 0.5b, all give similar output.

To reproduce: `./parallel -ngl 400 -np 3 -ns 10 -fa -m some-model.gguf`.

The broken batch sizes and backends I've tried:

| -np \<n\> | has corruption |
|-|-|
| 1, 2 | no |
| 3, 4, 5, 6, 7, 8 | yes |
| 9, 10, ..., 25 | no |
| > 25 | didn't test |

| backend | has corruption |
|-|-|
| cpu | no |
| cuda w/ gtx 1080 | yes |
| rocm w/ mi60 | no |

### Name and Version

$ ./parallel --version
version: 3138 (704a35b1)
built with cc (Debian 13.2.0-25) 13.2.0 for x86_64-linux-gnu

### What operating system are you seeing the problem on?

Linux

### Relevant log output

```shell
$ ./parallel -m /home/eric/mnt/models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf -ngl 400 -np 3 -ns 10 -fa -ngl 400
Log start
llama_model_loader: loaded meta data with 26 key-value pairs and 291 tensors from /home/eric/mnt/models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = Meta-Llama-3-8B-Instruct
llama_model_loader: - kv   2:                          llama.block_count u32              = 32
llama_model_loader: - kv   3:                       llama.context_length u32              = 8192
llama_model_loader: - kv   4:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   6:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv   7:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv   8:                       llama.rope.freq_base f32              = 500000.000000
llama_model_loader: - kv   9:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  10:                          general.file_type u32              = 15
llama_model_loader: - kv  11:                           llama.vocab_size u32              = 128256
llama_model_loader: - kv  12:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  13:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  14:                         tokenizer.ggml.pre str              = llama-bpe
llama_model_loader: - kv  15:                      tokenizer.ggml.tokens arr[str,128256]  = ["!", "\"", "#", "$", "%", "&", "'", ...
llama_model_loader: - kv  16:                  tokenizer.ggml.token_type arr[i32,128256]  = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...
llama_model_loader: - kv  17:                      tokenizer.ggml.merges arr[str,280147]  = ["Ġ Ġ", "Ġ ĠĠĠ", "ĠĠ ĠĠ", "...
llama_model_loader: - kv  18:                tokenizer.ggml.bos_token_id u32              = 128000
llama_model_loader: - kv  19:                tokenizer.ggml.eos_token_id u32              = 128001
llama_model_loader: - kv  20:                    tokenizer.chat_template str              = {% set loop_messages = messages %}{% ...
llama_model_loader: - kv  21:               general.quantization_version u32              = 2
llama_model_loader: - kv  22:                      quantize.imatrix.file str              = /models/Meta-Llama-3-8B-Instruct-GGUF...
llama_model_loader: - kv  23:                   quantize.imatrix.dataset str              = /training_data/groups_merged.txt
llama_model_loader: - kv  24:             quantize.imatrix.entries_count i32              = 224
llama_model_loader: - kv  25:              quantize.imatrix.chunks_count i32              = 88
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type q4_K:  193 tensors
llama_model_loader: - type q6_K:   33 tensors
llm_load_vocab: special tokens cache size = 256
llm_load_vocab: token to piece cache size = 0.8000 MB
llm_load_print_meta: format           = GGUF V3 (latest)
llm_load_print_meta: arch             = llama
llm_load_print_meta: vocab type       = BPE
llm_load_print_meta: n_vocab          = 128256
llm_load_print_meta: n_merges         = 280147
llm_load_print_meta: n_ctx_train      = 8192
llm_load_print_meta: n_embd           = 4096
llm_load_print_meta: n_head           = 32
llm_load_print_meta: n_head_kv        = 8
llm_load_print_meta: n_layer          = 32
llm_load_print_meta: n_rot            = 128
llm_load_print_meta: n_embd_head_k    = 128
llm_load_print_meta: n_embd_head_v    = 128
llm_load_print_meta: n_gqa            = 4
llm_load_print_meta: n_embd_k_gqa     = 1024
llm_load_print_meta: n_embd_v_gqa     = 1024
llm_load_print_meta: f_norm_eps       = 0.0e+00
llm_load_print_meta: f_norm_rms_eps   = 1.0e-05
llm_load_print_meta: f_clamp_kqv      = 0.0e+00
llm_load_print_meta: f_max_alibi_bias = 0.0e+00
llm_load_print_meta: f_logit_scale    = 0.0e+00
llm_load_print_meta: n_ff             = 14336
llm_load_print_meta: n_expert         = 0
llm_load_print_meta: n_expert_used    = 0
llm_load_print_meta: causal attn      = 1
llm_load_print_meta: pooling type     = 0
llm_load_print_meta: rope type        = 0
llm_load_print_meta: rope scaling     = linear
llm_load_print_meta: freq_base_train  = 500000.0
llm_load_print_meta: freq_scale_train = 1
llm_load_print_meta: n_ctx_orig_yarn  = 8192
llm_load_print_meta: rope_finetuned   = unknown
llm_load_print_meta: ssm_d_conv       = 0
llm_load_print_meta: ssm_d_inner      = 0
llm_load_print_meta: ssm_d_state      = 0
llm_load_print_meta: ssm_dt_rank      = 0
llm_load_print_meta: model type       = 8B
llm_load_print_meta: model ftype      = Q4_K - Medium
llm_load_print_meta: model params     = 8.03 B
llm_load_print_meta: model size       = 4.58 GiB (4.89 BPW)
llm_load_print_meta: general.name     = Meta-Llama-3-8B-Instruct
llm_load_print_meta: BOS token        = 128000 '<|begin_of_text|>'
llm_load_print_meta: EOS token        = 128001 '<|end_of_text|>'
llm_load_print_meta: LF token         = 128 'Ä'
llm_load_print_meta: EOT token        = 128009 '<|eot_id|>'
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:   no
ggml_cuda_init: CUDA_USE_TENSOR_CORES: yes
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce GTX 1080, compute capability 6.1, VMM: yes
llm_load_tensors: ggml ctx size =    0.30 MiB
llm_load_tensors: offloading 32 repeating layers to GPU
llm_load_tensors: offloading non-repeating layers to GPU
llm_load_tensors: offloaded 33/33 layers to GPU
llm_load_tensors:        CPU buffer size =   281.81 MiB
llm_load_tensors:      CUDA0 buffer size =  4403.49 MiB
........................................................................................
llama_new_context_with_model: n_ctx      = 8192
llama_new_context_with_model: n_batch    = 2048
llama_new_context_with_model: n_ubatch   = 512
llama_new_context_with_model: flash_attn = 1
llama_new_context_with_model: freq_base  = 500000.0
llama_new_context_with_model: freq_scale = 1
llama_kv_cache_init:      CUDA0 KV buffer size =  1024.00 MiB
llama_new_context_with_model: KV self size  = 1024.00 MiB, K (f16):  512.00 MiB, V (f16):  512.00 MiB
llama_new_context_with_model:  CUDA_Host  output buffer size =     1.96 MiB
llama_new_context_with_model:      CUDA0 compute buffer size =   258.50 MiB
llama_new_context_with_model:  CUDA_Host compute buffer size =    24.01 MiB
llama_new_context_with_model: graph nodes  = 903
llama_new_context_with_model: graph splits = 2

No new questions so proceed with build-in defaults.


main: Simulating parallel requests from clients:
main: n_parallel = 3, n_sequences = 10, cont_batching = 1, system tokens = 259

main: Evaluating the system prompt ...

Processing requests ...

main: clearing the KV cache
Client   0, seq    0, started decoding ...
Client   1, seq    1, started decoding ...
Client   2, seq    2, started decoding ...
Client   0, seq   0/ 10, prompt    9 t, response    3 t, time  0.59 s, speed 20.40 t/s, cache miss 0
Input:    What is the meaning of life?
Response: This
P

Client   0, seq    3, started decoding ...
Client   0, seq   3/ 10, prompt   20 t, response    3 t, time  0.18 s, speed 125.59 t/s, cache miss 0
Input:    Are you familiar with the Special Theory of Relativity and can you explain it to me?
Response: Yes

Client   0, seq    4, started decoding ...
Client   0, seq   4/ 10, prompt    9 t, response    4 t, time  0.21 s, speed 60.84 t/s, cache miss 0
Input:    Recommend some interesting books to read.
Response: I- S

Client   0, seq    5, started decoding ...
Client   0, seq   5/ 10, prompt   20 t, response    3 t, time  0.18 s, speed 127.55 t/s, cache miss 0
Input:    Are you familiar with the Special Theory of Relativity and can you explain it to me?
Response: Yes

...

Client   0, seq    6, started decoding ...
Client   0, seq   6/ 10, prompt   15 t, response    3 t, time  0.18 s, speed 100.19 t/s, cache miss 0
Input:    If you could have any superpower, what would it be?
Response: I
...S

Client   0, seq    7, started decoding ...
Client   1, seq   1/ 10, prompt   12 t, response   23 t, time  1.49 s, speed 23.56 t/s, cache miss 0
Input:    What is the best way to cook a steak?
Response: The best way to cook a steak is to grill it. Here's a simple recipe to achieve a perfectly cooked steak:

Client   1, seq    8, started decoding ...
Client   0, seq   7/ 10, prompt   12 t, response    5 t, time  0.27 s, speed 64.05 t/s, cache miss 0
Input:    I want to learn how to play the piano.
Response: The [...] ... T

Client   0, seq    9, started decoding ...
Client   0, seq   9/ 10, prompt    9 t, response    3 t, time  0.17 s, speed 69.66 t/s, cache miss 0
Input:    Recommend some interesting books to read.
Response: I…H

Client   2, seq   2/ 10, prompt    9 t, response  100 t, time  4.26 s, speed 25.59 t/s, cache miss 0
Input:    What is the meaning of life?
Response: This is a question that has puzzled philosophers and scientists for centuries. There is no one definitive answer, but many people believe that the meaning of life is to find happiness, fulfillment, and purpose through our experiences and relationships. Others believe that the meaning of life is to seek out knowledge and understanding, and to use our talents and abilities to make a positive impact on the world. Ultimately, the meaning of life is a personal and subjective question, and it is up to each individual to find their own answer.

Client   1, seq   8/ 10, prompt   13 t, response   84 t, time  3.03 s, speed 32.00 t/s, cache miss 0
Input:    What is the best way to learn a new language?
Response: There are several ways to learn a new language, but the best way is through immersion. Immersion is the process of surrounding yourself with the language you want to learn, such as by living in a country where the language is spoken, listening to music and watching movies in the language, and speaking with native speakers. You can also use language learning apps, such as Duolingo, and practice with a language exchange partner.

main: clearing the KV cache

run parameters as at 2024-06-12 05:33:13

main: n_parallel = 3, n_sequences = 10, cont_batching = 1, system tokens = 259
External prompt file: used built-in defaults
Model and path used:  /home/eric/mnt/models/Meta-Llama-3-8B-Instruct-Q4_K_M.gguf

Total prompt tokens:    128, speed: 28.28 t/s
Total gen tokens:       231, speed: 51.04 t/s
Total speed (AVG):           speed: 79.31 t/s
Cache misses:             0


llama_print_timings:        load time =     950.53 ms
llama_print_timings:      sample time =      23.11 ms /   241 runs   (    0.10 ms per token, 10427.03 tokens per second)
llama_print_timings: prompt eval time =    4095.15 ms /   610 tokens (    6.71 ms per token,   148.96 tokens per second)
llama_print_timings:        eval time =     253.14 ms /     8 runs   (   31.64 ms per token,    31.60 tokens per second)
llama_print_timings:       total time =    4526.35 ms /   618 tokens
```

**Comment by ggerganov** [comment]
Does it depend on the quant type? Does it happen with F16 models?

**Comment by nytopop** [comment]
I've tried Q4_K_M, Q8_0, and fp16. They all show the same issue.

**Comment by JohannesGaessler** [comment]
I believe to have found the issue, please confirm whether this fix works: https://github.com/ggerganov/llama.cpp/pull/7904

**Comment by nytopop** [comment]
Probably limited to cards without tensor cores, then. Looking at `ggml-cuda/fattn.cu`: https://github.com/ggerganov/llama.cpp/blob/a9cae48003dfc4fe95b8f5c81682fc6e63425235/ggml-cuda/fattn.cu#L316-L323

I see corrupt output specifically when `ggml_cuda_flash_attn_ext_vec_f32` is chosen. The `tile` variant seems fine.

On the rocm machine i'm testing on, `ggml_cuda_flash_attn_ext_vec_f16` is used and that produces normal output.

**Comment by nytopop** [comment]
> I believe to have found the issue, please confirm whether this fix works: #7904

Does seem to fix it, yes.
